### PR TITLE
feat(esbuild): plugin for ESBuild & Stylable

### PR DESCRIPTION
**Pull request by barak007**

### CODE REVIEW

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+                resolveNamespace: resolveNamespaceNode,
+                resolveModule: configFromFile?.config?.defaultConfig?.resolveModule,
+            },
+            build
+        );
+
+        const stylable = new Stylable(stConfig);
+
+        build.onStart(() => {
+            stylable.initCache();
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:106]: add test for hmr

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        });
+
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:108]: add comment explaining the general flow of loading stylable files

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+                contents: wrapWithDepthMarkers(res.meta.targetAst!.toString(), cssDepth),
+                loader: 'css',
+            };
+        });
+
+        build.onLoad({ filter: /.*/, namespace: namespaces.jsModule }, (args) => {
+            const res = stylable.transform(args.path);
+            const { cssDepth = 0, deepDependencies } = res.meta.transformCssDepth!;
+            const imports = [];
+            const unusedImports = [];
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:148]: `unusedImports` are collected but not used - use in optimizations?

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+                path: stylable.resolvePath(
+                    args.resolveDir,
+                    args.path.replace(namespaces.nativeCss + `:`, '')
+                ),
+                namespace: cssInjection === 'css' ? namespaces.nativeCss : namespaces.jsModule,
+            };
+        });
+
+        build.onLoad({ filter: /.*/, namespace: namespaces.nativeCss }, (args) => {
+            const res = stylable.transform(args.path);
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:135]: is the transform happening twice? isn't it on the `pluginData`?

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+
+function buildUsageMapping(metafile: Metafile, stylable: Stylable) {
+    const usageMapping: Record<string, boolean> = {};
+    const usages: Record<string, Set<string>> = {};
+    for (const [key] of Object.entries(metafile.inputs)) {
+        if (key.startsWith(namespaces.jsModule)) {
+            const meta =
+                stylable.fileProcessor.cache[key.replace(namespaces.jsModule + ':', '')].value;
+            if (!meta) {
+                throw new Error('meta not found');
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:283]: make message more clear

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            }
+            usages[meta.namespace] ||= new Set();
+            usages[meta.namespace].add(key);
+            usageMapping[meta.namespace] = true;
+        }
+    }
+    for (const [namespace, usage] of Object.entries(usages)) {
+        if (usage.size > 1) {
+            console.warn(`namespace: ${namespace} used in:\n${[...usage].join('\n')}`);
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:292]:
```suggestion
            console.warn(`namespace: ${namespace} used in more then one stylesheet:\n${[...usage].join('\n')}`);
```

Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:292]: add test to check duplicate namespace (pass warn to esbuild?)

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+            const { 1: esbuildComment, 2: depth, 3: css } = args;
+            extracted.push({ depth: parseInt(depth, 10), css: (esbuildComment || '') + css });
+            return '';
+        }
+    );
+
+    const sorted = sortModulesByDepth(
+        extracted,
+        (m) => m.depth,
+        () => /*TODO: should we sort by id like in webpack? */ '',
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:377]: pass id here

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+            .join('')
+    );
+}
+
+function removeUnusedComponents(
+    css: string,
+    stylable: Stylable,
+    usageMapping: Record<string, boolean>
+) {
+    usageMapping['deep'] = false;
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:396]: remove and use `unusedImports` to here

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+        usageMapping,
+        {
+            classes: {},
+            containers: {},
+            keyframes: {},
+            vars: {},
+            layers: {},
+            stVars: {},
+        },
+        { global: true } // TODO: handle globals!!!!! get usage from meta
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:410]: handle globals...

**Review by idoros — CHANGES_REQUESTED** [review]

File: packages/cli/src/build-single-file.ts [hunk 10]
```diff
@@ -149,29 +148,8 @@ export function buildSingleFile({
     moduleFormats.forEach(([format, ext]) => {
         outputLogs.push(`${format} module`);
 
-        const moduleCssImports = [];
-
+        const moduleCssImports = collectImportsWithSideEffects(res, stylable, ext);
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/cli/src/build-single-file.ts:151]: it was but now not. the refactor was nice I kept it.

File: packages/rollup-plugin/src/index.ts [hunk 11]
```diff
@@ -100,31 +100,32 @@ export function stylableRollupPlugin({
...
-                stConfig.projectRoot,
-                typeof stcConfig === 'string' ? stcConfig : undefined,
-                fs
-            );
 
             if (stylable) {
                 clearRequireCache();
                 stylable.initCache();
             } else {
+                const stConfig = stylableConfig({
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/rollup-plugin/src/index.ts:114]: it's not related. just when I looked wanted an example I saw it and fixed it.

File: packages/esbuild/README.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+```ts
+interface ESBuildOptions {
+    /**
+     * Determine the way css is injected to the document
+     * js - every js module contains the css and inject it independently
+     * css - emit bundled css asset to injected via link
+     */
+    cssInjection?: 'js' | 'css';
+    /**
+     * Config how error and warning reported to webpack by stylable
```
[the remark this answers is not in the document]
Comment (barak007) [reply packages/esbuild/README.md:36]: done

File: packages/esbuild/test/e2e/simple-case/build.js [hunk 13]
```diff
@@ -10,25 +11,21 @@ module.exports.cssBundleProd = (build, options) => {
 };
 
 function run(cssInjection, build, options) {
+    const outdir = cssInjection === 'css' ? 'dist-bundle' : 'dist';
```
Comment (idoros) [inline packages/esbuild/test/e2e/simple-case/build.js:14]: maybe change the out-dir name to reflect the setting that it relates to? something like `dist-css/css-in-js`?

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 14]
```diff
@@ -94,80 +96,76 @@ export const stylablePlugin = (initialPluginOptions: ESBuildOptions = {}): Plugi
...
+                ),
+                namespace: namespaces.unused,
+            };
+        });
+
+        /**
+         * handle css/stylable files imported via other stylable files
+         */
+        build.onResolve({ filter: /\.css$/, namespace: namespaces.jsModule }, (args) => {
+            // self import of css injection
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:131]: stylable file generated JavaScript module import self CSS source

File: packages/esbuild/src/stylable-esbuild-plugin.ts [hunk 15]
```diff
@@ -94,80 +96,76 @@ export const stylablePlugin = (initialPluginOptions: ESBuildOptions = {}): Plugi
...
                 namespace: cssInjection === 'css' ? namespaces.nativeCss : namespaces.jsModule,
+                pluginData: args.pluginData,
             };
         });
 
-        build.onLoad({ filter: /.*/, namespace: namespaces.nativeCss }, (args) => {
-            const res = stylable.transform(args.path);
-            const cssDepth = res.meta.transformCssDepth!.cssDepth;
+        /**
+         * handle all initial stylable requests from
```
Comment (idoros) [inline packages/esbuild/src/stylable-esbuild-plugin.ts:151]: from JavaScript?

File: packages/esbuild/test/e2e/simple-case/build.js (hunk 13 above)
Comment (barak007) [reply packages/esbuild/test/e2e/simple-case/build.js:14]: I don't think it's necessary the test is already overloaded with terms so `css-in-js` might not be better than `dev`.

File: packages/esbuild/test/e2e/rebuild/rebuild.spec.ts [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+    it('should pick up rebuild', async () => {
+        const { context, read, write } = await tk.build({
+            project: 'rebuild',
+            tmp: true,
+        });
+        const css1 = read('dist/index.js');
+        expect(css1, 'initial color').to.includes('color: red');
+        await context.watch();
+        await sleep(2222);
```
Comment (idoros) [inline packages/esbuild/test/e2e/rebuild/rebuild.spec.ts:18]: is there a way to wait for build to be done without sleeping for 4.5s?

Comment (barak007) [reply packages/esbuild/test/e2e/rebuild/rebuild.spec.ts:18]: not that I know. he state the the watch loop is at most 2s. so I put little more

**Review by idoros — APPROVED** [review]
